### PR TITLE
Added a proof that all polymorphic functions A -> A are id

### PR DESCRIPTION
**Pull request by jozefg**
This is a surprisingly simple proof. The main trick is using the subset of  A contracted around the argument.

### CODE REVIEW

**Comment by jonsterling** [comment]
Ah, yes! I remember this one from OPLSS. Very nice!

Can we rename the development to `polymorphic-id-unique` or something though?

**Comment by jozefg** [comment]
@jonsterling Oh yeah that's a bit clearer, renamed!

**Comment by jonsterling** [comment]
Great! Would you also add this to the makefile? thanks!

**Comment by jonsterling** [comment]
Oh, and I meant to rename the file too :)

File: example/id.jonprl [hunk 1]
```diff
@@ -0,0 +1,41 @@
+Theorem polymorphic-is-unique : [
```
Comment (jonsterling) [inline example/id.jonprl]: `s/is/id`

**Comment by jozefg** [comment]
Oh good point, done and done :)

File: Makefile [hunk 2]
```diff
@@ -17,6 +17,7 @@ test:
 	bin/jonprl example/russell.jonprl
 	bin/jonprl example/container.jonprl
 	bin/jonprl example/list.jonprl
+	bin/jonprl example/id.jonprl
```
Comment (jonsterling) [inline Makefile]: need this one too

**Comment by jozefg** [comment]
@jonsterling Sorry for the delay, just got back home. I think I fixed everything now :)

**Comment by jonsterling** [comment]
Awesome, thanks again! :)

**Comment by ghost** [comment]
This is really interesting. Does this sort of approach generalize to more complicated situations?

**Comment by jonsterling** [comment]
We asked Bickford about that—there are a number of parametricity results that are not provable in nuprl (i can't remember specifics, but i recall feeling relieved about some of them)...

But many nice examples do go through.

**Comment by ithaca-markb** [comment]
If we write (intersection A: universe_i. ...) as (alltype A. ...) then
some other parametricity results that work  are:
f: alltype A. A -> Bool    <=>  exists b:Bool.  f = \x.b
 (this also works for alltype A. A -> T, as long as T is a "value-type" (does not contain bottom) and has Krary's "mono" property   (all x,y:T.  (x sqle y) => x = y)  )   -- where sqle is Howe's computational approximation relation. So T = int,  int \* int,  int list,  etc. all work.

I can also prove that a polymorphic choice  g: alltype A. A -> A -> A has to
be either fst =  \x y. x  or  snd = \x y. y   when restricted to non-bottom  terms.
But there are two other (I think the only other) polymorphic choice functions:
fst' = \x y.  let z = eval y in x
snd' = \x y. let z = eval x in y
(so fst' x y = x unless y=bottom, and fst x bottom = bottom)
Showing that  fst'  has type alltype A. A -> A -> A   seems to require classical logic to claim that every type A either contains bottom or does not contain bottom. But use of classical logic is consistent with Nuprl when proving a membership judgement, so we can't rule out fst' and snd'.

In system F , where everything terminates, fst' = fst and snd' = snd, so this distinction does not arise.

Another example, from a paper by Harper & Krary,  is a "polymorphic select" of type
h:  alltype A.   (A List+) -> A    ---  where List+  is non-nil lists
We expect that the only thing h(L) can do is  choose an i < length(L) and return  L[i]
so h(L) = L[h [0;1;...;length(L)-1]]   and  h(map(f,L)) = f(h(L)).
But again, we can only prove this for types that do not contain bottom, because
in general h can first evaluate some of the elements of L.

**Comment by jozefg** [comment]
@ithaca-markb Thanks for the explanation! I may have to have a crack at proving some of those things with jonprl..

**Comment by ithaca-markb** [comment]
Hey! I was wrong about the functions of type alltype A. A -> A -> A in Nuprl.
Those two functions  fst' = \x y. let z = eval y in x  and  snd' = \x y. let z = eval x in y
do not have type alltype A. A -> A -> A  after all.  The reason is that (for example) 
fst' 0 bottom = bottom  and fst' 0 1 = 0, but we can create a quotient type A in which
(bottom = 1 in A) but  not(0=1 in A), so if fst' had type A->A->A then we would get bottom=0 in A
and hence (0=1 in A) and so False.

In fact, I was able to prove that for all f: alltype A. A -> A -> A,  
(all x,y:Base. f x y = x)  or (all x, y: Base. f x y = y)

(my proof was a bit tricky, I needed to consider several different quotient types).

So, probably, the Harper & Krary result about h:  alltype A. (A List+) -> A
(that h(map(f,L)) = f (h L) )  is also true in Nuprl.

**Pull request merged**